config/internal/configsource/configsource.go

```diff
@@ -0,0 +1,315 @@
+				code: errCfgSrcBeginSession,
+			}
+		}
+
+		// The scope of usage of the cfgSrc is the whole function so it is fine to defer
+		// the EndSession call to the function exit.
+		defer cfgSrc.EndSession(ctx)
+	}
+
+	srcCfg := v
+	var dstCfg *viper.Viper
+
+	var err error
+	done := false
+	for i := -1; i <= int(params.MaxRecursionDepth) && !done; i++ {
+		dstCfg = config.NewViper()
+		done, err = applyConfigSources(ctx, srcCfg, dstCfg, params.ConfigSources)
+		if err != nil {
+			return nil, err
+		}
+		srcCfg = dstCfg
+	}
+
+	if !done {
+		return nil, &cfgSrcError{
+			msg:  "config source recursion chain is too deep, couldn't fully expand the configuration",
+			code: errCfgSrcChainTooLong,
+		}
+	}
+
+	return dstCfg, nil
+}
+
+// cfgSrcError private error type used to accurate identify the type of error in tests.
+type cfgSrcError struct {
+	msg  string          // human readable error message.
+	code cfgSrcErrorCode // internal error code.
+}
+
+func (e *cfgSrcError) Error() string {
+	return e.msg
+}
+
+// These are errors that can be returned by ApplyConfigSources function.
+// Note that error codes are not part public API, they are for unit testing only.
+type cfgSrcErrorCode int
+
+const (
+	_ cfgSrcErrorCode = iota // skip 0, start errors codes from 1.
+	errCfgSrcChainTooLong
+	errOnlyMapAtRootLevel
+	errCfgSrcBeginSession
+	errCfgSrcNotFound
+	errCfgSrcApply
+)
+
+func applyConfigSources(ctx context.Context, srcCfg, dstCfg *viper.Viper, cfgSources map[string]component.ConfigSource) (bool, error) {
+	// Expand any item env vars in the config, do it every time so env vars
+	// added on previous pass are also handled.
+	expandEnvConfig(srcCfg)
+
+	done := true
+	appliedTags := make(map[string]struct{})
+
+	// It is possible to have config sources injection depending on other config sources:
+	//
+	// $lower_cfgsrc:
+	//   a: "just an example"
+	//   b:
+	//     $deeper_cfgsrc:
+	//       c: true
+	//
+	// By injecting the deepest ones first they can be resolved in the proper order.
+	// See function deepestConfigSourcesFirst for more info.
+	//
+	// TODO: Bug when lower injection happening after deeper but without the inject values.
+	allKeys := srcCfg.AllKeys()
+	sort.Slice(allKeys, deepestConfigSourcesFirst(allKeys))
+
+	// Inspect all key from original configuration and set the proper value on the destination config.
+	for _, k := range allKeys {
+		dstKey, cfgSrcName, paramsKey := extractCfgSrcInvocation(k)
+		if cfgSrcName == "" {
+			// Nothing to apply take the key and value as it is.
+			dstCfg.Set(k, srcCfg.Get(k))
+			continue
+		}
+
+		if _, ok := appliedTags[dstKey]; ok {
+			// Already applied this key. If the config source has multiple parameters
+			// there will be multiple keys for the same application.
+			continue
+		}
+
+		appliedTags[dstKey] = struct{}{}
+		cfgSrc, ok := cfgSources[cfgSrcName]
+		if !ok {
+			return false, &cfgSrcError{
+				msg:  fmt.Sprintf("config source %s not found", cfgSrcName),
```

Comment:
```suggestion
				msg:  fmt.Sprintf("config source %q not found", cfgSrcName),
```

Reply:
Looking at this and other usages of this object, I kinda wish that the message was a structured log entry at debug level, instead of having a single string with all the information.

Reply:
This is keeping in line with the `config` package and its related packages: they return errors instead of directly logging messages.